server/jetstream_test.go

```diff
@@ -9922,6 +9922,64 @@ func TestJetStreamPullConsumerMaxWaitingOfOne(t *testing.T) {
 	}
 }
 
+func TestJetStreamPullConsumerMaxWaitingUpdate(t *testing.T) {
+	s := RunBasicJetStreamServer()
+	if config := s.JetStreamConfig(); config != nil {
+		defer removeDir(t, config.StoreDir)
+	}
+	defer s.Shutdown()
+
+	nc, js := jsClientConnect(t, s)
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{Name: "TEST", Subjects: []string{"TEST.A"}})
+	require_NoError(t, err)
+
+	_, err = js.AddConsumer("TEST", &nats.ConsumerConfig{
+		Durable:   "dur",
+		AckPolicy: nats.AckExplicitPolicy,
+	})
+	require_NoError(t, err)
+
+	// Update from default max waiting (512) to 1
+	_, err = js.UpdateConsumer("TEST", &nats.ConsumerConfig{
+		Durable:    "dur",
+		AckPolicy:  nats.AckExplicitPolicy,
+		MaxWaiting: 1,
+	})
+	require_NoError(t, err)
+
+	sub, err := js.PullSubscribe("", "dur", nats.Bind("TEST", "dur"))
+	require_NoError(t, err)
+
+	// Test max waiting requests
+	errch := make(chan error, 2)
+
+	go func() {
+		_, err := sub.Fetch(1, nats.MaxWait(time.Second))
+		errch <- err
+	}()
+
+	go func() {
+		_, err := sub.Fetch(1, nats.MaxWait(time.Second))
+		errch <- err
+	}()
+
+	var ok bool
+	err = <-errch
+	if strings.Contains(err.Error(), "Exceeded MaxWaiting") {
+		ok = true
+	}
+	err = <-errch
```

Comment:
Since you have MaxWaiting == 1 and have only 2 concurrent requests, only 1 should fail because of Exceeded MaxWaiting, not the 2. So I would recommend that you setup the one that you know is going to fail due to a timeout (so bigger timeout) and the second should fail due to Exceeded MaxWaiting. Let's also try to keep the MaxWait as small as possible (but without falling into flapping territory).

Reply:
Thanks, good point. Pushed a new commit, let me know if this is sufficient.

Reply:
And for context in case @derekcollison didn't mention, this is intended to fail right now to show case the `MaxWaiting` option is not be reflected on consumer update. It still maintains its original `waitQueue` size of 512 after update.